datafusion/expr/src/logical_plan/ddl.rs

```diff
@@ -288,8 +290,234 @@ impl PartialOrd for CreateExternalTable {
+            table_partition_cols: vec![],
+            if_not_exists: false,
+            temporary: false,
+            definition: None,
+            order_exprs: vec![],
+            unbounded: false,
+            options: HashMap::new(),
+            constraints: Constraints::empty(),
+            column_defaults: HashMap::new(),
+        }
+    }
+
+    pub fn name(mut self, name: TableReference) -> Self {
+        self.name = Some(name);
+        self
+    }
+
+    pub fn schema(mut self, schema: DFSchemaRef) -> Self {
+        self.schema = Some(schema);
+        self
+    }
+
+    pub fn location(mut self, location: String) -> Self {
+        self.location = Some(location);
+        self
+    }
+
+    pub fn file_type(mut self, file_type: String) -> Self {
+        self.file_type = Some(file_type);
+        self
+    }
+
+    pub fn table_partition_cols(mut self, table_partition_cols: Vec<String>) -> Self {
+        self.table_partition_cols = table_partition_cols;
+        self
+    }
+
+    pub fn if_not_exists(mut self, if_not_exists: bool) -> Self {
+        self.if_not_exists = if_not_exists;
+        self
+    }
+
+    pub fn temporary(mut self, temporary: bool) -> Self {
+        self.temporary = temporary;
+        self
+    }
+
+    pub fn definition(mut self, definition: Option<String>) -> Self {
+        self.definition = definition;
+        self
+    }
+
+    pub fn order_exprs(mut self, order_exprs: Vec<Vec<Sort>>) -> Self {
+        self.order_exprs = order_exprs;
+        self
+    }
+
+    pub fn unbounded(mut self, unbounded: bool) -> Self {
+        self.unbounded = unbounded;
+        self
+    }
+
+    pub fn options(mut self, options: HashMap<String, String>) -> Self {
+        self.options = options;
+        self
+    }
+
+    pub fn constraints(mut self, constraints: Constraints) -> Self {
+        self.constraints = constraints;
+        self
+    }
+
+    pub fn column_defaults(mut self, column_defaults: HashMap<String, Expr>) -> Self {
+        self.column_defaults = column_defaults;
+        self
+    }
+
+    pub fn build(self) -> Result<CreateExternalTable> {
+        CreateExternalTable::new(CreateExternalTableFields {
+            name: self.name.expect("name is required"),
+            schema: self.schema.expect("schema is required"),
+            location: self.location.expect("location is required"),
+            file_type: self.file_type.expect("file_type is required"),
+            table_partition_cols: self.table_partition_cols,
+            if_not_exists: self.if_not_exists,
+            temporary: self.temporary,
+            definition: self.definition,
+            order_exprs: self.order_exprs,
+            unbounded: self.unbounded,
+            options: self.options,
+            constraints: self.constraints,
+            column_defaults: self.column_defaults,
+        })
+    }
+}
+
 /// Creates an in memory table.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Hash)]
+#[non_exhaustive]
```

Comment:
why add this?

Reply:
To disallow public direct construction. If we allow public direct construction, we can't enforce invariants in the factory method, since it may or may not get called.